acorn/src/regexp.js

```diff
@@ -920,6 +981,249 @@ pp.regexp_eatClassEscape = function(state) {
   )
 }
 
+// https://tc39.es/ecma262/#prod-ClassSetExpression
+// https://tc39.es/ecma262/#prod-ClassUnion
+// https://tc39.es/ecma262/#prod-ClassIntersection
+// https://tc39.es/ecma262/#prod-ClassSubtraction
+pp.regexp_classSetExpression = function(state) {
+  let nextMayContainStrings = false
+  if (this.regexp_eatClassSetRange(state))
+    ;
+  else if (this.regexp_eatClassSetOperand(state)) {
+    nextMayContainStrings = state.lastMayContainStrings
+    let pos = state.pos
+    const start = pos
+    // https://tc39.es/ecma262/#prod-ClassIntersection
+    while (
+      state.eat(0x26 /* & */) && state.eat(0x26 /* & */) &&
```

Comment:
Could we maybe add some lookahead or multi-character eat mechanism to `RegExpValidationState`? This repeated side-effecting with `eat` only to restore the position after seems both hard to read and error-prone.